quickwit-actors/src/mailbox.rs

```diff
@@ -0,0 +1,271 @@
+        match self {
+            Command::Pause => write!(f, "Pause"),
+            Command::Stop(_) => write!(f, "Stop"),
+            Command::Start => write!(f, "Start"),
+            Command::Observe(_) => write!(f, "Observe"),
+            Command::ScheduledMessage(_) => write!(f, "ScheduleMsg"),
+        }
+    }
+}
+
+impl<Message: fmt::Debug> fmt::Debug for Mailbox<Message> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Mailbox({})", self.actor_instance_name())
+    }
+}
+
+impl<Message> Hash for Mailbox<Message> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.id.hash(state)
+    }
+}
+
+impl<Message> PartialEq for Mailbox<Message> {
+    fn eq(&self, other: &Self) -> bool {
+        self.id.eq(&other.id)
+    }
+}
+
+impl<Message> Eq for Mailbox<Message> {}
+
+impl<Message> Mailbox<Message> {
+    pub fn actor_instance_name(&self) -> String {
+        format!("{}:{}", self.actor_name, self.id)
+    }
+
+    pub(crate) async fn send_actor_message(
+        &self,
+        msg: ActorMessage<Message>,
+    ) -> Result<(), SendError> {
+        self.sender.send_async(msg).await?;
+        Ok(())
+    }
+
+    /// Send a message to the actor synchronously.
+    ///
+    /// SendError is returned if the user is already terminated.
+    ///
+    /// (See also [Self::send_blocking()])
+    pub(crate) async fn send_message(&self, msg: Message) -> Result<(), SendError> {
+        self.send_actor_message(ActorMessage::Message(msg)).await
+    }
+
+    /// Send a message to the actor in a blocking fashion.
+    /// When possible, prefer using [Self::send()].
+    pub(crate) fn send_message_blocking(&self, msg: Message) -> Result<(), SendError> {
+        self.sender.send(ActorMessage::Message(msg))?;
+        Ok(())
+    }
+
+    pub fn send_command_blocking(&self, command: Command<Message>) -> Result<(), SendError> {
+        self.command_sender.send(command)?;
+        Ok(())
+    }
+
+    pub async fn send_command(&self, command: Command<Message>) -> Result<(), SendError> {
+        self.command_sender.send_async(command).await?;
+        Ok(())
+    }
+}
+
+pub struct Inbox<Message> {
+    pub rx: flume::Receiver<ActorMessage<Message>>,
+    command_rx: flume::Receiver<Command<Message>>,
+}
+
+#[derive(Debug)]
+pub enum ReceptionResult<M> {
+    Command(Command<M>),
+    Message(M),
+    None,
+    Disconnect, // was disconnect from either the command or the regular mailbox.
+}
+
+impl<Message: fmt::Debug> Inbox<Message> {
+    fn get_command_if_available(&self) -> Option<ReceptionResult<Message>> {
+        match self.command_rx.try_recv() {
+            Ok(command) => Some(ReceptionResult::Command(command)),
+            Err(TryRecvError::Disconnected) => Some(ReceptionResult::Disconnect),
+            Err(TryRecvError::Empty) => None,
+        }
+    }
+
+    pub(crate) async fn try_recv_msg(&self, message_enabled: bool) -> ReceptionResult<Message> {
```

Comment:
I don't understand the concept of message enabled.

Reply:
Yes, it's not obvious. My understanding is that it avoids making a if else when the actor is paused by passing a boolean to `try_recv_msg`. So whatever the actor state, we are just calling this function.
Personnally I would prefer to make the `if paused return None else try_recv_msg` but I'm ok to keep it like this.

Reply:
I am not sure this helped but I move the logic outside.

The idea is that when the actor is paused, we stop processing regular messages, but we keep processing command.
One glitchy spec is that we also process scheduled message. (I'm not sure this is a good thing....)